lib/request-handler.js

```diff
@@ -63,94 +63,97 @@ module.exports = function processRequest (options, request, response) {
     };
 
     let shouldWriteHead = true;
-    let index = 0;
 
-    contextPromise.then((context) => {
+    const contentLengthStream = new ContentLengthStream((contentLength) => {
+        this.emit('end', request, contentLength);
+    });
 
-        const contentLengthStream = new ContentLengthStream((contentLength) => {
-            this.emit('end', request, contentLength);
-        });
+    const handleError = (err) => {
+        this.emit('error', request, err);
+        if (shouldWriteHead) {
+            shouldWriteHead = false;
+            let statusCode = 500;
+            if (err.code === TEMPLATE_NOT_FOUND) {
+                statusCode = 404;
+            }
 
-        const resultStream = new StringifierStream((tag) => {
-            const { placeholder, name, } = tag;
-            if (placeholder === 'pipe') {
-                return pipeDefinition(pipeInstanceName);
+            response.writeHead(statusCode, responseHeaders);
+            // To render with custom error template
+            if (typeof err.presentable === 'string') {
+                response.end(`${err.presentable}`);
+            } else {
+                response.end();
             }
+        } else {
+            contentLengthStream.end();
+        }
+    };
+
+    Promise.all([
+        templatePromise,
+        contextPromise
+    ]).then(([ parsedTemplate, context ]) => {
+        const blockStream = buildBlock(request, Object.assign({
+            index: 0,
```

Comment:
don't allow useland code to update index at any point.

Reply:
Wow, you're right. Thanks for catching!